Documentation/design-docs/hw-intrinsics.md

```diff
@@ -0,0 +1,81 @@
+Implementation of Hardware Intrinsics in CoreCLR
+================================================
+This document describes the implementation of hardware intrinsics in CoreCLR.
+For information about how the intrinsic APIs are designed, proposed and approved,
+see https://github.com/dotnet/designs/blob/master/accepted/platform-intrinsics.md.
+
+In discussing the hardware intrinsics, we refer to the target platform, such as X86 or Arm64, as the "platform" and each set of extensions that are implemented as a unit (e.g. AVX2 on X64 or Simd on Arm64) as an "ISA".
+
+There is a design document for the Arm64 intrinsics: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/arm64-intrinsics.md. It should be updated to reflect current (and ongoing) progress.
+
+## Overview
+
+The reference assemblies for the hardware intrinsics live in corefx, but all of the implementation is in the coreclr repo:
+
+* The C# implementation lives in coreclr/src/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that usually invoke themselves recursively (see Indirect invocation support below).
+
+## C# Implementation
+
+The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types. The vector types are considered platform-agnostic, though not all platforms define intrinsics on all of these types. They are:
+
+* `Vector64<T>` - A 64-bit vector of type `T`. For example, a `Vector64<int>` would hold two 32-bit integers.
```

Comment:
One of the things that comes up in Arm64 is that a vector must be two or more.  So `Vector64<double>` is not a `Vector<T>`

Reply:
Oh, weird. Perhaps I knew that before, but had forgotten. Was there a fundamental reason for that?

Reply:
I realize that it seems rather pointless, but one could imagine a scenario where I'd like to implement something over `Vector64<T>` and would like it to "just work" for the relevant types, and not have to special-case `double`.

Reply:
> Was there a fundamental reason for that?

There aren't any instructions that operate this way from what I recall. One could conceivable consider it to be the same as only exposing scalar operations for the given type, but there was some JIT support that needed to happen here last I looked.

Reply:
This was briefly called out here: https://github.com/dotnet/coreclr/pull/21198#discussion_r236537125

Basically ARM64 supports `1D` for some, but not all instructions; so the JIT needs to know when it should use `V.1D` and when it should emit the scalar equivalent. Also the JIT doesn't currently allow `TYP_SIMD8` for `double/int64/uint64` and that would also need updating.

Reply:
I am not sure how the Vector<T> type was implemented.  I would guess the Vector<double> just works but fall back to software emulation.
For the SIMD intrinsics, I think the Vector64<double> will PNSE.  These could philosophically be supported in some(most) circumstances, but they would have to be special cased to use the normal double form. I think since they map to a different instruction we didn't really consider it.